lib/bamboo/template.ex

```diff
@@ -0,0 +1,271 @@
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def welcome_email do
+          new_email()
+          |> render(:welcome)
+        end
+      end
+
+  _Make assigns available to a template._
+
+      defmodule MyApp.Email do
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def welcome_email(user) do
+          new_email()
+          |> assign(:user, user)
+          |> render(:welcome)
+        end
+      end
+
+  _Make assigns available to a template during render call._
+
+      defmodule MyApp.Email do
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def welcome_email(user) do
+          new_email()
+          |> put_html_layout({MyApp.Email.LayoutView, "email.html"})
+          |> render(:welcome, user: user)
+        end
+      end
+
+  _Render an email by passing the template string to render._
+
+      defmodule MyApp.Email do
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def html_email do
+          new_email()
+          |> render("html_email.html")
+        end
+
+        def text_email do
+          new_email
+          |> render("text_email.text")
+        end
+      end
+
+  ## HTML Layout Example
+
+      # lib/my_app/email.ex
+      defmodule MyApp.Email do
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def welcome_email(person) do
+          base_email()
+          |> to(person)
+          |> subject("Welcome to MyApp")
+          |> assign(:person, person)
+          |> render(:welcome_email)
+        end
+
+        defp base_email do
+          new_email()
+          |> from("Rob Ot<robot@changelog.com>")
+          |> put_header("Reply-To", "editors@changelog.com")
+          # This will use the "email.html.eex" file as a layout when rendering html emails.
+          # Plain text emails will not use a layout unless you use `put_text_layout`
+          |> put_html_layout({MyApp.Email.LayoutView, "email.html"})
+        end
+      end
+
+      # lib/my_app/email/views/layout_view.ex
+      defmodule MyApp.Email.LayoutView do
+        use Bamboo.View, path: "/lib/my_app/email/templates/layout"
+      end
+
+      # lib/my_app/email/templates/layout/email.html.eex
+      <html>
+        <body>
+          <%= @inner_content %>
+        </body>
+      </html>
+
+      # lib/my_app/email/views/account_view.ex
+      defmodule MyApp.Email.AccountView do
+        use Bamboo.View, path: "/lib/my_app/email/templates/account"
+      end
+
+      # lib/my_app/email/templates/account/welcome_email.html.eex
+      # This will be rendered within a layout because `put_html_layout` was used.
+      <p>Welcome <%= @person.name %></p>
+
+      # lib/my_app/email/templates/account/welcome_email.text.eex
+      # This will not be rendered within a layout because `put_text_layout` was not used.
+      Welcome <%= @person.name %>
+  """
+
+  import Bamboo.Email, only: [put_private: 3]
+
+  defmacro __using__(view: view_module) do
```

Comment:
What would you think about accepting layouts as `use` options? To take one of your examples from above:
This:

```elixir
defmodule MyApp.Email do
  use Bamboo.Template, view: MyApp.Email.AccountView
  def welcome_email do
    new_email()
    |> put_text_layout({MyApp.Email.LayoutView, "email.text"})
    |> put_html_layout({MyApp.Email.LayoutView, "email.html"})
    |> render(:welcome) # Pass atom to render html AND plain text templates
  end
end
```

could be used this way:

```elixir
defmodule MyApp.Email do
  use Bamboo.Template,
    view: MyApp.Email.AccountView,
    layout_html: {MyApp.Email.LayoutView, "email.html"},
    layout_text: {MyApp.Email.LayoutView, "email.text"}
  
  def welcome_email do
    # Pass atom to render html AND plain text templates
    render(new_email(), :welcome) 
  end
end
```

I'm thinking that if you define multiple email functions within the module, the layout is set for all of them removing the need to call the layout functions directly each time.

Thoughts?

Reply:
Yeah, I like this idea. 👍  Let me see what we'd need to do in order to be able to do that but also be able to override the layout. I don't think it'd be a big change.

Reply:
I updated this in 84a679435cd47086e85414f8c10fa98ea5d95855 if you'd like to take a look. I like how things turned out.